src_new/util.idhub.js

```diff
@@ -0,0 +1,438 @@
+	var referenceForNesting = convertedObject;
+	for (var i = 0; i < sourceArray.length - 1; i++) {
+		if (!referenceForNesting[sourceArray[i]]) {
+			referenceForNesting[sourceArray[i]] = {};
+		}
+		referenceForNesting = referenceForNesting[sourceArray[i]];
+	}
+	referenceForNesting[sourceArray[sourceArray.length - 1]] = valueOfLastNode;
+	return convertedObject;
+};
+
+exports.getNestedObjectFromString = function (sourceObject, separator, key, value) {
+	var splitParams = key.split(separator);
+	if (splitParams.length == 1) {
+		sourceObject[key] = value;
+	} else {
+		sourceObject = refThis.getNestedObjectFromArray(sourceObject, splitParams, value);
+	}
+	return sourceObject;
+};
+
+
+exports.initLiveRampAts = function (params) {
+	function addATS() {
+		var atsScript = document.createElement("script");
+		if (params.params.cssSelectors && params.params.cssSelectors.length > 0) {
+			params.params.cssSelectors = params.params.cssSelectors.split(",");
+		}
+		atsScript.onload = function () {
+			window.ats.start({
+				"placementID": params.params.pid,
+				"storageType": params.params.storageType,
+				"detectionType": params.params.detectionType,
+				"urlParameter": params.params.urlParameter,
+				"cssSelectors": params.params.cssSelectors, // ["input[type=text]", "input[type=email]"],
+				"logging": params.params.logging //"error"
+			});
+		};
+		atsScript.src = "https://ats.rlcdn.com/ats.js";
+		document.body.appendChild(atsScript);
+	}
+	window.addEventListener("load", function () {
+		setTimeout(addATS, 1000);
+	});
+};
+
+exports.updateAdUnits = function (adUnits) {
+	if (refThis.isArray(adUnits)) {
+		adUnits.forEach(function (adUnit) {
+			adUnit.bids.forEach(function (bid) {
+				refThis.updateUserIds(bid);
+			});
+		});
+	} else if (!refThis.isEmptyObject(adUnits)) {
+		adUnits.bids.forEach(function (bid) {
+			refThis.updateUserIds(bid);
+		});
+	}
+};
+
+exports.updateUserIds = function (bid) {
+	// refThis.idsAppendedToAdUnits =true;
+	if (refThis.isUndefined(bid.userId)) {
+		bid["userId"] = refThis.getUserIds();
+	} else if (bid.userId) {
+		/* istanbul ignore next */
+		bid.userId = Object.assign(bid.userId, refThis.getUserIds());
+	}
+	if (refThis.isUndefined(bid.userIdAsEids)) {
+		bid["userIdAsEids"] = refThis.getUserIdsAsEids();
+	} else if (refThis.isArray(bid.userIdAsEids)) {
+		var idsPresent = new Set();
+		var ids = refThis.getUserIdsAsEids().concat(bid.userIdAsEids);
+		if (refThis.isArray(ids) && ids.length > 0) {
+			ids = ids.filter(function (id) {
+				if (id.source) {
+					if (idsPresent.has(id.source)) {
+						return false;
+					}
+					idsPresent.add(id.source);
+				}
+				return true;
+
+			});
+		}
+		bid.userIdAsEids = ids;
+	}
+};
+
+exports.applyDataTypeChangesIfApplicable = function(params) {
+	var value;
+	if(params.name in CONSTANTS.SPECIAL_CASE_ID_PARTNERS) {
+		for(partnerName in CONSTANTS.SPECIAL_CASE_ID_PARTNERS) {
+			for(key in CONSTANTS.SPECIAL_CASE_ID_PARTNERS[partnerName]) {
```

Comment:
![Codacy](https://codacy.pubmatic.com:443/assets/images/favicon.png) Issue found: ['partnerName' is not defined.](https://codacy.pubmatic.com:443/manual/dattaprasad.mundada/OpenWrap/pullRequest?prid=15746)